### ISSUE TITLE
Masked Softmax produces differing softmax weights depending on the amount of masking on some architectures.

### ISSUE THREAD

**Issue by villmow** [pytorch/pytorch#115417]
### 🐛 Describe the bug

To compute a masked softmax we fill the masked values with the lowest number possible. ([See the implementation in the transformers repository](https://github.com/huggingface/transformers/blob/56be5e80e6cd5264012eb9ea84bd589233a503d9/src/transformers/modeling_utils.py#L859)). I noticed that will lead to different softmax scores on some machines. 

In the example script below I create two tensors with differing length (186 and 256) filled with the lowest number possible. In these two tensors I input the same 55 values (for which I want to compute the softmax scores) at different positions. 

**I would expect that both tensors contain the same softmax scores!** 

On a linux system this works as expected and the softmax scores are the same. However on mac and windows there is a difference. I tested it on different systems and attached the prints below.

Is this wanted behaviour? When stacked in multiple transformer layers this leads to quite some difference in the transformer output, depending on the amount and position of padding in the input tensors.

```python
import torch
import platform
import sys

print("Python Version:", sys.version)
print("Torch version: ", torch.__version__)
print("System: ", platform.system())
print("Node: ", platform.node())
print("Release: ", platform.release())
print("Version: ", platform.version())
print("Machine: ", platform.machine())
print("Processor: ", platform.processor())
print("Platform: ", platform.platform())

# some scores tensor
# fmt: off
a = [
    -3.9325406551361084, 0.9592550992965698, 1.1994309425354004, 0.6425036787986755, 1.2429596185684204, 0.5248308777809143, 1.0555568933486938, 0.3594433069229126, 0.7644527554512024, 0.7945529818534851, 0.9265341758728027, 0.2066846489906311, 0.5986100435256958, 0.7194857001304626, 0.6189113855361938, 0.5716232061386108, -0.277407705783844, 0.4133310317993164, -0.06546378135681152, 0.163144052028656, 0.3343402147293091, -0.277407705783844, 0.31046062707901, 0.07031236588954926, 0.1911880075931549, -0.33945488929748535, 0.0906137079000473, -0.01491139829158783, -0.33945488929748535, -0.5645690560340881, 0.1442396193742752, -0.5645690560340881, -0.3931613862514496, -0.8763489723205566, -0.27457353472709656, -0.288480669260025, -0.24146756529808044, -0.07116357982158661, -0.8763489723205566, -0.39812904596328735, -0.268454372882843, -0.8763489723205566, 0.07237496972084045, -0.8763489723205566, -0.3931613862514496, -0.8763489723205566, -0.5904549360275269, -0.6043620705604553, -0.5573489665985107, -0.5843358039855957, -0.5843358039855957, -0.36273038387298584, -0.9802864789962769, -0.5629925727844238, -0.43925708532333374,
]
# fmt: on
a = torch.tensor(a)

# initialize two tensors with negative infinity
x = torch.full((186,), torch.finfo(torch.float32).min)
y = torch.full((256,), torch.finfo(torch.float32).min)

# fill the first 55 elements of x with the scores
x[:55] = a
# fill some middle elements of y with the same scores
y[63 : 63 + 55] = a

# check if the first 55 elements of x and the middle elements of y are equal
x1 = x[:55]
y1 = y[63 : 63 + 55]
print("x1 == y1:", torch.equal(x1, y1))

# compute softmax over both tensors
attn_x = torch.nn.functional.softmax(x, dim=-1)
attn_y = torch.nn.functional.softmax(y, dim=-1)

attn_x1 = attn_x[:55]
attn_y1 = attn_y[63 : 63 + 55]

print("DIFF = ", (attn_x1 - attn_y1).abs().sum())
```

Outputs:

```
################################
Python Version: 3.9.16 (main, Jan 11 2023, 16:05:54)
[GCC 11.2.0]
Torch version:  1.12.1
System:  Linux
Node:  thinkpad
Release:  5.11.0-051100-generic
Version:  #202102142330 SMP Sun Feb 14 23:33:21 UTC 2021
Machine:  x86_64
Processor:  x86_64
Platform:  Linux-5.11.0-051100-generic-x86_64-with-glibc2.31

x1 == y1: True
DIFF =  tensor(0.)
################################
Python Version: 3.10.0 | packaged by conda-forge | (default, Nov 20 2021, 02:24:10) [GCC 9.4.0]
Torch version:  2.1.1+cu121
System:  Linux
Node:  thinkpad
Release:  5.11.0-051100-generic
Version:  #202102142330 SMP Sun Feb 14 23:33:21 UTC 2021
Machine:  x86_64
Processor:  x86_64
Platform:  Linux-5.11.0-051100-generic-x86_64-with-glibc2.31

x1 == y1: True
DIFF =  tensor(0.)	
################################
Python Version: 3.10.12 (main, Jul  5 2023, 15:02:25) [Clang 14.0.6 ]
Torch version:  2.0.1
System:  Darwin
Node:  Person1-MacBook-Pro.local
Release:  23.1.0
Version:  Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:45 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6020
Machine:  arm64
Processor:  arm
Platform:  macOS-14.1.2-arm64-arm-64bit

x1 == y1: True
DIFF =  tensor(1.3228e-07)
################################
Python Version: 3.9.6 (default, Aug 11 2023, 19:44:50)
[Clang 15.0.0 (clang-1500.0.40.1)]
Torch version: 1.12.1
System: Darwin
Node: Person2-MacBook-Pro.local
Release: 23.1.0
Version: Darwin Kernel Version 23.1.0: Mon Oct 9 21:27:27 PDT 2023; root:xnu-10002.41.9~6/RELEASE_X86_64
Machine: x86_64
Processor: i386
Platform: macOS-14.1.2-x86_64-i386-64bit

x1 == y1: True
DIFF = tensor(1.3041e-07)
################################
Python Version: 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:14:58) [MSC v.1929 64 bit (AMD64)]
Torch version:  2.0.0+cpu
System:  Windows
Node:  DESKTOP-LJOMQKN
Release:  10
Version:  10.0.19045
Machine:  AMD64
Processor:  Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
Platform:  Windows-10-10.0.19045-SP0

x1 == y1: True
DIFF =  tensor(1.3228e-07)
```

### Versions

PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.1.2 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.12 (main, Jul  5 2023, 15:02:25) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.1.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] pytorch-lightning==2.0.7
[pip3] pytorch-metric-learning==2.3.0
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchdata==0.6.1
[pip3] torchmetrics==1.1.1
[pip3] torchvision==0.15.2a0
[conda] faiss-cpu                 1.7.4           py3.10_h34215f8_0_cpu    pytorch
[conda] libfaiss                  1.7.4            hcb8d3e5_0_cpu    pytorch
[conda] numpy                     1.25.2          py310h3b2db8e_0  
[conda] numpy-base                1.25.2          py310ha9811e2_0  
[conda] pytorch                   2.0.1                  py3.10_0    pytorch
[conda] pytorch-lightning         2.0.7                    pypi_0    pypi
[conda] pytorch-metric-learning   2.3.0                    pypi_0    pypi
[conda] torchaudio                2.0.2                 py310_cpu    pytorch
[conda] torchdata                 0.6.1                    pypi_0    pypi
[conda] torchmetrics              1.1.1                    pypi_0    pypi
[conda] torchvision               0.15.2          cpu_py310h31aa045_0

**Comment by colesbury** [comment]
Floating point addition and multiplication are not associative. If you change the order elements you may get slightly different results even if mathematically they should be the same. Each platform also uses different kernels, so the results will not necessarily be exactly identical between platforms.

**Comment by lezcano** [comment]
Relevant docs: https://pytorch.org/docs/master/notes/numerical_accuracy.html#batched-computations-or-slice-computations
Closing as expected.
